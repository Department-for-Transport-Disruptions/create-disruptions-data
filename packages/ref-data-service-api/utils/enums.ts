export enum RefVehicleMode {
    bus = "bus",
    tram = "tram",
    coach = "coach",
    ferry = "ferry",
    underground = "underground",
    rail = "rail",
    metro = "metro",
    blank = "",
}
