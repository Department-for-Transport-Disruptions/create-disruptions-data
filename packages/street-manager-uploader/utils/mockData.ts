const sqsMessage = {
    permitReferenceNumber: "TSR1591199404915-01",
    highwayAuthority: "CITY OF WESTMINSTER",
    highwayAuthoritySwaCode: 5990,
    worksLocationCoordinates: "LINESTRING(501251.53 222574.64,501305.92 222506.65)",
    streetName: "HIGH STREET NORTH",
    areaName: "LONDON",
    workCategory: "Standard",
    trafficManagementType: "Road closure",
    proposedStartDateTime: "2020-06-10T00:00:00.000Z",
    proposedEndDateTime: "2020-06-12T00:00:00.000Z",
    actualStartDateTime: null,
    actualEndDateTime: null,
    workStatus: "Works planned",
    usrn: "8401426",
    activityType: "Remedial works",
    worksLocationType: "Cycleway, Footpath",
    isTrafficSensitive: "Yes",
    permitStatus: "permit_modification_request",
    town: "LONDON",
    currentTrafficManagementType: "Multi-way signals",
    currentTrafficManagementTypeUpdateDate: null,
    lastUpdatedDateTime: "2020-06-04T08:00:00.000Z",
};

export const mockSqsEvent = {
    Records: [
        {
            messageId: "id",
            receiptHandle: "test",
            body: JSON.stringify(sqsMessage),
            md5OfBody: "test",
            eventSource: "test",
            eventSourceARN: "test",
            awsRegion: "test",
            attributes: {
                ApproximateReceiveCount: "test",
                SentTimestamp: "test",
                SenderId: "test",
                ApproximateFirstReceiveTimestamp: "test",
            },
            messageAttributes: {
                message: { dataType: "String" },
            },
        },
    ],
};
