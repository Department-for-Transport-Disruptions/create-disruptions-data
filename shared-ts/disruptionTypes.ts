import { z } from "zod";
import {
    consequenceOperatorsSchema,
    consequenceSchema,
    disruptionInfoSchema,
    disruptionSchema,
    networkConsequenceSchema,
    operatorConsequenceSchema,
    serviceSchema,
    servicesConsequenceSchema,
    stopSchema,
    stopsConsequenceSchema,
    validitySchema,
} from "./disruptionTypes.zod";

export type Disruption = z.infer<typeof disruptionSchema>;
export type DisruptionInfo = z.infer<typeof disruptionInfoSchema>;

export type Validity = z.infer<typeof validitySchema>;

export type Consequence = z.infer<typeof consequenceSchema>;
export type ConsequenceOperators = z.infer<typeof consequenceOperatorsSchema>;
export type Stop = z.infer<typeof stopSchema>;
export type Service = z.infer<typeof serviceSchema>;

export type NetworkConsequence = z.infer<typeof networkConsequenceSchema>;
export type OperatorConsequence = z.infer<typeof operatorConsequenceSchema>;
export type StopsConsequence = z.infer<typeof stopsConsequenceSchema>;
export type ServicesConsequence = z.infer<typeof servicesConsequenceSchema>;

export const routesSchema = z.object({
    inbound: z.array(stopSchema.partial()),
    outbound: z.array(stopSchema.partial()),
});

export type Routes = z.infer<typeof routesSchema>;
