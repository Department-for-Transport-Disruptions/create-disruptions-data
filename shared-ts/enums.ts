export enum SourceType {
    directReport = "directReport",
    email = "email",
    phone = "phone",
    fax = "fax",
    post = "post",
    feed = "feed",
    radio = "radio",
    tv = "tv",
    web = "web",
    pager = "pager",
    text = "text",
    other = "other",
}

export enum Progress {
    draft = "draft",
    open = "open",
    published = "published",
    closing = "closing",
    closed = "closed",
    pendingApproval = "pendingApproval",
    editPendingApproval = "editPendingApproval",
    draftPendingApproval = "draftPendingApproval",
    rejected = "rejected",
}

export enum MiscellaneousReason {
    accident = "accident",
    securityAlert = "securityAlert",
    congestion = "congestion",
    roadClosed = "roadClosed",
    incident = "incident",
    routeDiversion = "routeDiversion",
    unknown = "unknown",
    vandalism = "vandalism",
    overcrowded = "overcrowded",
    operatorCeasedTrading = "operatorCeasedTrading",
    vegetation = "vegetation",
    roadworks = "roadworks",
    specialEvent = "specialEvent",
    insufficientDemand = "insufficientDemand",
}

export enum PersonnelReason {
    unknown = "unknown",
    staffSickness = "staffSickness",
    staffInjury = "staffInjury",
    contractorStaffInjury = "contractorStaffInjury",
    staffAbsence = "staffAbsence",
    staffInWrongPlace = "staffInWrongPlace",
    staffShortage = "staffShortage",
    industrialAction = "industrialAction",
    unofficialIndustrialAction = "unofficialIndustrialAction",
    workToRule = "workToRule",
    undefinedPersonnelProblem = "undefinedPersonnelProblem",
}

export enum EquipmentReason {
    unknown = "unknown",
    pointsFailure = "pointsFailure",
    signalProblem = "signalProblem",
    trainWarningSystemProblem = "trainWarningSystemProblem",
    trackCircuitProblem = "trackCircuitProblem",
    signalFailure = "signalFailure",
    derailment = "derailment",
    engineFailure = "engineFailure",
    tractionFailure = "tractionFailure",
    breakDown = "breakDown",
    technicalProblem = "technicalProblem",
    brokenRail = "brokenRail",
    poorRailConditions = "poorRailConditions",
    wheelImpactLoad = "wheelImpactLoad",
    lackOfOperationalStock = "lackOfOperationalStock",
    defectiveFireAlarmEquipment = "defectiveFireAlarmEquipment",
    defectivePlatformEdgeDoors = "defectivePlatformEdgeDoors",
    defectiveCctv = "defectiveCctv",
    defectivePublicAnnouncementSystem = "defectivePublicAnnouncementSystem",
    ticketingSystemNotAvailable = "ticketingSystemNotAvailable",
    repairWork = "repairWork",
    constructionWork = "constructionWork",
    maintenanceWork = "maintenanceWork",
    emergencyEngineeringWork = "emergencyEngineeringWork",
    lateFinishToEngineeringWork = "lateFinishToEngineeringWork",
    powerProblem = "powerProblem",
    fuelProblem = "fuelProblem",
    swingBridgeFailure = "swingBridgeFailure",
    escalatorFailure = "escalatorFailure",
    liftFailure = "liftFailure",
    gangwayProblem = "gangwayProblem",
    closedForMaintenance = "closedForMaintenance",
    fuelShortage = "fuelShortage",
    deicingWork = "deicingWork",
    wheelProblem = "wheelProblem",
    luggageCarouselProblem = "luggageCarouselProblem",
    undefinedEquipmentProblem = "undefinedEquipmentProblem",
}

export enum EnvironmentReason {
    unknown = "unknown",
    fog = "fog",
    roughSea = "roughSea",
    heavySnowFall = "heavySnowFall",
    driftingSnow = "driftingSnow",
    blizzardConditions = "blizzardConditions",
    heavyRain = "heavyRain",
    strongWinds = "strongWinds",
    stormConditions = "stormConditions",
    stormDamage = "stormDamage",
    tidalRestrictions = "tidalRestrictions",
    highTide = "highTide",
    lowTide = "lowTide",
    ice = "ice",
    frozen = "frozen",
    hail = "hail",
    sleet = "sleet",
    highTemperatures = "highTemperatures",
    flooding = "flooding",
    waterlogged = "waterlogged",
    lowWaterLevel = "lowWaterLevel",
    highWaterLevel = "highWaterLevel",
    fallenLeaves = "fallenLeaves",
    fallenTree = "fallenTree",
    landslide = "landslide",
    undefinedEnvironmentalProblem = "undefinedEnvironmentalProblem",
    lightningStrike = "lightningStrike",
    sewerOverflow = "sewerOverflow",
    grassFire = "grassFire",
}

export enum DayType {
    monday = "monday",
    tuesday = "tuesday",
    wednesday = "wednesday",
    thursday = "thursday",
    friday = "friday",
    saturday = "saturday",
    sunday = "sunday",
}

export enum VehicleMode {
    bus = "bus",
    tram = "tram",
    ferryService = "ferryService",
    rail = "rail",
    underground = "underground",
}

export enum Severity {
    unknown = "unknown",
    normal = "normal",
    verySlight = "verySlight",
    slight = "slight",
    severe = "severe",
    verySevere = "verySevere",
}

export enum UserGroups {
    systemAdmins = "system-admins",
    orgAdmins = "org-admins",
    orgPublishers = "org-publishers",
    orgStaff = "org-staff",
    operators = "operators",
}

export enum PublishStatus {
    draft = "DRAFT",
    published = "PUBLISHED",
    editing = "EDITING",
    pendingApproval = "PENDING_APPROVAL",
    rejected = "REJECTED",
    editPendingApproval = "EDIT_PENDING_APPROVAL",
    pendingAndEditing = "PENDING_EDITING",
}

export enum SocialMediaPostStatus {
    pending = "Pending",
    successful = "Successful",
    rejected = "Rejected",
}

export enum Datasource {
    tnds = "tnds",
    bods = "bods",
}

export enum Modes {
    bus = "bus",
    coach = "coach",
    tram = "tram",
    ferry = "ferry",
    rail = "rail",
    underground = "underground",
    metro = "metro",
}
