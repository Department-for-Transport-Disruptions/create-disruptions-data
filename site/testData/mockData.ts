/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { mockRequest, mockResponse } from "mock-req-res";
import { NextPageContext } from "next";
import React from "react";
import { Mock, vi } from "vitest";
import { COOKIE_ID_TOKEN, COOKIES_POLICY_COOKIE } from "../constants";

export interface GetMockContextInput {
    session?: { [key: string]: any };
    cookies?: any;
    body?: any;
    url?: any;
    uuid?: any;
    mockWriteHeadFn?: Mock<any, any>;
    mockEndFn?: Mock<any, any>;
    isLoggedin?: boolean;
    query?: any;
}

export const getMockRequestAndResponse = ({
    cookieValues = {},
    body = null,
    mockWriteHeadFn = vi.fn(),
    mockEndFn = vi.fn(),
    requestHeaders = {},
    isLoggedin = true,
    url = null,
    query = null,
}: GetMockRequestAndResponse = {}): { req: any; res: any } => {
    const res = mockResponse({ writeHead: mockWriteHeadFn, end: mockEndFn });

    const {
        idToken = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJjdXN0b206bm9jIjoiVEVTVCIsImVtYWlsIjoidGVzdEBleGFtcGxlLmNvbSIsImp0aSI6Ijg1MmQ1MTVlLTU5YWUtNDllZi1iMTA5LTI4YTRhNzk3YWFkNSIsImlhdCI6MTU5Mjk4NzMwNywiZXhwIjoxNTkyOTkwOTA3fQ.DFdxnpdhykDONOMeZMNeMUFpCHZ-hQ3UXczq_Qh0IAI",
        cookiePolicy = null,
    } = cookieValues;

    const defaultSession = {
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        destroy: (): void => {},
    };

    let cookieString = "";

    cookieString += isLoggedin ? `${COOKIE_ID_TOKEN}=${idToken as string};` : "";

    cookieString += cookiePolicy ? `${COOKIES_POLICY_COOKIE}=${encodeURI(JSON.stringify(cookiePolicy))}` : "";

    const req = mockRequest({
        connection: {
            encrypted: true,
        },
        url,
        headers: {
            host: "localhost:3000",
            cookie: cookieString,
            origin: "localhost:3000",
            ...requestHeaders,
        },
        cookies: cookieValues,
        session: { ...defaultSession },
    });

    if (body) {
        req.body = body;
    }
    if (query) {
        req.query = query;
    }

    return { req, res };
};

export const getMockContext = ({
    cookies = {},
    body = null,
    uuid = {},
    mockWriteHeadFn = vi.fn(),
    mockEndFn = vi.fn(),
    isLoggedin = true,
    url = null,
    query = "",
}: GetMockContextInput = {}): NextPageContext => {
    const { req, res } = getMockRequestAndResponse({
        cookieValues: cookies,
        body,
        uuid,
        mockWriteHeadFn,
        mockEndFn,
        requestHeaders: {},
        isLoggedin,
        url,
    });

    const ctx: NextPageContext = {
        res,
        req,
        pathname: "",
        query,
        // eslint-disable-next-line react/display-name
        AppTree: () => React.createElement("div"),
    };

    return ctx;
};

export interface GetMockRequestAndResponse {
    cookieValues?: any;
    body?: any;
    uuid?: any;
    mockWriteHeadFn?: Mock<any, any>;
    mockEndFn?: Mock<any, any>;
    requestHeaders?: any;
    isLoggedin?: boolean;
    url?: any;
    query?: any;
}
