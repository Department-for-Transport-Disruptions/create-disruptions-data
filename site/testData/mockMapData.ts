export const mockMarkerData = [
    {
        atcoCode: "1800SB12581",
        commonName: "Cheapside",
        indicator: "Stop CL",
        longitude: -2.244551,
        latitude: 53.48096,
        bearing: "W",
        stopType: "BCT",
        busStopType: "MKD",
    },
    {
        atcoCode: "1800SB05801",
        commonName: "Cheapside",
        indicator: "Stop CO",
        longitude: -2.243361,
        latitude: 53.48097,
        bearing: "E",
        stopType: "BCT",
        busStopType: "MKD",
    },
    {
        atcoCode: "1800SB05751",
        commonName: "Manchester Town Hall",
        indicator: "Stop SC",
        longitude: -2.244697,
        latitude: 53.480024,
        bearing: "SE",
        stopType: "BCT",
        busStopType: "MKD",
    },
];
