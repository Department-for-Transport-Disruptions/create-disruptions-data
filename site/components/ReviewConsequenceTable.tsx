import { Consequence, Disruption } from "@create-disruptions-data/shared-ts/disruptionTypes";
import { MAX_CONSEQUENCES } from "@create-disruptions-data/shared-ts/disruptionTypes.zod";
import Link from "next/link";
import { ReactElement, ReactNode } from "react";
import Table, { CellProps } from "./form/Table";
import {
    CANCELLATIONS_FEATURE_FLAG,
    CONSEQUENCE_TYPES,
    CREATE_CONSEQUENCE_NETWORK_PATH,
    CREATE_CONSEQUENCE_OPERATOR_PATH,
    CREATE_CONSEQUENCE_SERVICES_PATH,
    CREATE_CONSEQUENCE_STOPS_PATH,
    DISRUPTION_DETAIL_PAGE_PATH,
    REVIEW_DISRUPTION_PAGE_PATH,
    TYPE_OF_CONSEQUENCE_PAGE_PATH,
    VEHICLE_MODES,
} from "../constants";
import { getDisplayByValue, splitCamelCaseToString } from "../utils";

const getConsequenceUrl = (type: Consequence["consequenceType"]) => {
    switch (type) {
        case "networkWide":
            return CREATE_CONSEQUENCE_NETWORK_PATH;
        case "operatorWide":
            return CREATE_CONSEQUENCE_OPERATOR_PATH;
        case "stops":
            return CREATE_CONSEQUENCE_STOPS_PATH;
        case "services":
            return CREATE_CONSEQUENCE_SERVICES_PATH;
    }
};

export const createChangeLink = (
    key: string,
    href: string,
    disruptionId: string,
    index?: number,
    includePreviousPage?: boolean,
    isDisruptionDetail?: boolean,
    isTemplate?: boolean,
) => {
    return (
        <Link
            key={key}
            className="govuk-link"
            href={{
                pathname: `${href}/${disruptionId}${index !== undefined ? `/${index}` : ""}`,
                query: includePreviousPage
                    ? {
                          return: isDisruptionDetail ? DISRUPTION_DETAIL_PAGE_PATH : REVIEW_DISRUPTION_PAGE_PATH,
                          ...(isTemplate ? { template: isTemplate.toString() } : {}),
                      }
                    : isTemplate
                    ? { template: isTemplate.toString() }
                    : null,
            }}
        >
            Change
        </Link>
    );
};

const getRows = (
    consequence: Consequence,
    disruption: Disruption,
    isEditingAllowed: boolean,
    isDisruptionDetail?: boolean,
    isTemplate?: boolean,
) => {
    const rows: { header?: string | ReactNode; cells: CellProps[] }[] = [
        {
            header: "Consequence type",
            cells: [
                {
                    value: getDisplayByValue(
                        CONSEQUENCE_TYPES(CANCELLATIONS_FEATURE_FLAG),
                        consequence.consequenceType,
                    ),
                    styles: {
                        width: "w-1/2",
                    },
                },
                {
                    value:
                        isEditingAllowed &&
                        createChangeLink(
                            "consequence-type",
                            TYPE_OF_CONSEQUENCE_PAGE_PATH,
                            disruption.disruptionId,
                            consequence.consequenceIndex,
                            true,
                            isDisruptionDetail,
                            isTemplate,
                        ),
                    styles: {
                        width: "w-1/10",
                    },
                },
            ],
        },
        {
            header: "Mode of transport",
            cells: [
                { value: getDisplayByValue(VEHICLE_MODES, consequence.vehicleMode) },
                {
                    value:
                        isEditingAllowed &&
                        createChangeLink(
                            "vehicle-mode",
                            getConsequenceUrl(consequence.consequenceType) || "",
                            disruption.disruptionId,
                            consequence.consequenceIndex,
                            true,
                            isDisruptionDetail,
                            isTemplate,
                        ),
                },
            ],
        },
    ];

    if (consequence.consequenceType === "services") {
        rows.push({
            header: "Service(s)",
            cells: [
                {
                    value: consequence.services
                        .map(
                            (service) =>
                                `${service.lineName} - ${service.origin} - ${service.destination} (${service.operatorShortName})`,
                        )
                        .join(", "),
                },
                {
                    value:
                        isEditingAllowed &&
                        createChangeLink(
                            "service",
                            getConsequenceUrl(consequence.consequenceType),
                            disruption.disruptionId,
                            consequence.consequenceIndex,
                            true,
                            isDisruptionDetail,
                            isTemplate,
                        ),
                },
            ],
        });
    }

    if (consequence.consequenceType === "services" || consequence.consequenceType === "stops") {
        rows.push({
            header: "Stops affected",
            cells: [
                {
                    value: consequence.stops
                        ? consequence.stops
                              .map((stop) =>
                                  stop.commonName && stop.indicator && stop.atcoCode
                                      ? `${stop.commonName} ${stop.indicator} ${stop.atcoCode}`
                                      : `${stop.commonName} ${stop.atcoCode}`,
                              )
                              .join(", ")
                        : "N/A",
                },
                {
                    value:
                        isEditingAllowed &&
                        createChangeLink(
                            "stops-affected",
                            getConsequenceUrl(consequence.consequenceType),
                            disruption.disruptionId,
                            consequence.consequenceIndex,
                            true,
                            isDisruptionDetail,
                            isTemplate,
                        ),
                },
            ],
        });
    }

    if (consequence.consequenceType === "operatorWide") {
        rows.push({
            header: "Operators affected",
            cells: [
                {
                    value: consequence.consequenceOperators
                        ? consequence.consequenceOperators.map((op) => op.operatorNoc).join(", ")
                        : "N/A",
                },
                {
                    value:
                        isEditingAllowed &&
                        createChangeLink(
                            "operators-affected",
                            getConsequenceUrl(consequence.consequenceType),
                            disruption.disruptionId,
                            consequence.consequenceIndex,
                            true,
                            isDisruptionDetail,
                            isTemplate,
                        ),
                },
            ],
        });
    }
    if (consequence.consequenceType === "networkWide") {
        rows.push({
            header: "Disruption Area",
            cells: [
                {
                    value: consequence.disruptionArea ? [...consequence.disruptionArea].join(", ") : "N/A",
                },
                {
                    value:
                        isEditingAllowed &&
                        createChangeLink(
                            "disruption-area",
                            getConsequenceUrl(consequence.consequenceType),
                            disruption.disruptionId,
                            consequence.consequenceIndex,
                            true,
                            isDisruptionDetail,
                            isTemplate,
                        ),
                },
            ],
        });
    }

    rows.push(
        {
            header: "Advice to display",
            cells: [
                {
                    value: consequence.description,
                },
                {
                    value:
                        isEditingAllowed &&
                        createChangeLink(
                            "advice-to-display",
                            getConsequenceUrl(consequence.consequenceType),
                            disruption.disruptionId,
                            consequence.consequenceIndex,
                            true,
                            isDisruptionDetail,
                            isTemplate,
                        ),
                },
            ],
        },
        {
            header: "Remove from journey planner",
            cells: [
                {
                    value: splitCamelCaseToString(consequence.removeFromJourneyPlanners),
                },
                {
                    value:
                        isEditingAllowed &&
                        createChangeLink(
                            "remove-from-journey-planners",
                            getConsequenceUrl(consequence.consequenceType),
                            disruption.disruptionId,
                            consequence.consequenceIndex,
                            true,
                            isDisruptionDetail,
                            isTemplate,
                        ),
                },
            ],
        },
        {
            header: "Disruption delay",
            cells: [
                {
                    value: consequence.disruptionDelay ? `${consequence.disruptionDelay} minutes` : "N/A",
                },
                {
                    value:
                        isEditingAllowed &&
                        createChangeLink(
                            "disruption-delay",
                            getConsequenceUrl(consequence.consequenceType),
                            disruption.disruptionId,
                            consequence.consequenceIndex,
                            true,
                            isDisruptionDetail,
                            isTemplate,
                        ),
                },
            ],
        },
    );
    return rows;
};

interface ReviewConsequenceTableProps {
    consequence: Consequence;
    disruption: Disruption;
    deleteActionHandler: (
        name: string,
        hiddenInputs: {
            name: string;
            value: string;
        }[],
    ) => void;
    isDisruptionDetail?: boolean;
    isTemplate?: boolean;
    isEditingAllowed: boolean;
}

const ReviewConsequenceTable = ({
    consequence,
    disruption,
    deleteActionHandler,
    isDisruptionDetail,
    isTemplate,
    isEditingAllowed,
}: ReviewConsequenceTableProps): ReactElement => {
    const hiddenInputs = [
        {
            name: "id",
            value: consequence.consequenceIndex.toString(),
        },
        {
            name: "disruptionId",
            value: disruption.disruptionId,
        },
    ];

    if (isDisruptionDetail) {
        hiddenInputs.push({
            name: "inEdit",
            value: "true",
        });
    }
    return (
        <>
            <Table rows={getRows(consequence, disruption, isEditingAllowed, isDisruptionDetail, isTemplate)} />
            {isEditingAllowed && (
                <>
                    <button
                        key={consequence.consequenceIndex}
                        className="govuk-button govuk-button--warning mt-4"
                        data-module="govuk-button"
                        onClick={(e) => {
                            e.preventDefault();
                            deleteActionHandler("consequence", hiddenInputs);
                        }}
                    >
                        Delete consequence
                    </button>
                    <button
                        key={`duplicate-${consequence.consequenceIndex}`}
                        className={`govuk-button govuk-button--secondary mt-4 ml-4${
                            disruption.consequences && disruption.consequences.length >= MAX_CONSEQUENCES
                                ? " pointer-events-none govuk-button--disabled"
                                : ""
                        }`}
                        data-module="govuk-button"
                        formAction={`/api/duplicate-consequence?consequenceId=${consequence.consequenceIndex}&return=${
                            isDisruptionDetail ? DISRUPTION_DETAIL_PAGE_PATH : REVIEW_DISRUPTION_PAGE_PATH
                        }${isTemplate ? "&template=true" : ""}`}
                    >
                        Duplicate consequence
                    </button>
                </>
            )}
        </>
    );
};

export default ReviewConsequenceTable;
