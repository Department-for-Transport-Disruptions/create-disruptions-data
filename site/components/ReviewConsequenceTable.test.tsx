import { Consequence, Disruption } from "@create-disruptions-data/shared-ts/disruptionTypes";
import { EnvironmentReason, PublishStatus, Severity, VehicleMode } from "@create-disruptions-data/shared-ts/enums";
import renderer from "react-test-renderer";
import { describe, expect, it, vi } from "vitest";
import { DEFAULT_ORG_ID } from "../testData/mockData";
import ReviewConsequenceTable from "./ReviewConsequenceTable";

const previousConsequencesInformation: Consequence = {
    vehicleMode: VehicleMode.bus,
    consequenceType: "networkWide",
    consequenceIndex: 0,
    disruptionId: "1",
    description: "The road is closed for the following reasons: Example, example, example, example",
    removeFromJourneyPlanners: "yes",
    disruptionDelay: "33",
    disruptionSeverity: Severity.severe,
};

const previousDisruptionInformation: Disruption = {
    disruptionType: "planned",
    disruptionId: "2",
    summary: "Road closure due to flooding and cattle on road and no sign of movement example example example etc etc",
    description:
        "Road closure due to flooding and cattle on road and no sign of movement example example example etc etc",
    associatedLink: "https://www.flooding.com",
    disruptionReason: EnvironmentReason.flooding,
    validity: [
        {
            disruptionStartDate: "13/01/2022",
            disruptionStartTime: "1200",
            disruptionEndDate: "14/01/2022",
            disruptionEndTime: "1400",
            disruptionNoEndDateTime: "",
        },
    ],
    publishStartDate: "13/01/2022",
    publishStartTime: "1300",
    publishEndDate: "13/01/2022",
    publishEndTime: "1400",
    disruptionStartDate: "15/01/2022",
    disruptionStartTime: "1200",
    disruptionEndDate: "17/01/2022",
    disruptionEndTime: "1400",
    disruptionNoEndDateTime: "",
    consequences: [previousConsequencesInformation],
    publishStatus: PublishStatus.draft,
    displayId: "8fg3ha",
    orgId: DEFAULT_ORG_ID,
    template: false,
};

describe("ReviewConsequenceTable", () => {
    it("should render the table with data", () => {
        const tree = renderer
            .create(
                <ReviewConsequenceTable
                    consequence={previousConsequencesInformation}
                    disruption={previousDisruptionInformation}
                    deleteActionHandler={vi.fn()}
                    isEditingAllowed={true}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
    it("should render the table with data without change links", () => {
        const tree = renderer
            .create(
                <ReviewConsequenceTable
                    consequence={previousConsequencesInformation}
                    disruption={previousDisruptionInformation}
                    deleteActionHandler={vi.fn()}
                    isEditingAllowed={false}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
