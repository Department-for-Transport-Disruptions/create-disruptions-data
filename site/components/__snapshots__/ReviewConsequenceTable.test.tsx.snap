// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ReviewConsequenceTable > should render the table with data 1`] = `
<DocumentFragment>
  <dl
    class="govuk-summary-list"
  >
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Consequence type
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        Network wide
      </dd>
      <dd
        class="govuk-summary-list__actions"
      >
        <ul
          class="govuk-summary-list__actions-list"
        >
          <li
            class="govuk-summary-list__actions-list-item"
          >
            <a
              class="govuk-link"
              href="/type-of-consequence/2/0?return=%2Freview-disruption"
            >
              Change
              <span
                class="govuk-visually-hidden"
              >
                Consequence type
              </span>
            </a>
          </li>
        </ul>
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Mode of transport
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        Bus
      </dd>
      <dd
        class="govuk-summary-list__actions"
      >
        <ul
          class="govuk-summary-list__actions-list"
        >
          <li
            class="govuk-summary-list__actions-list-item"
          >
            <a
              class="govuk-link"
              href="/create-consequence-network/2/0?return=%2Freview-disruption"
            >
              Change
              <span
                class="govuk-visually-hidden"
              >
                Mode of transport
              </span>
            </a>
          </li>
        </ul>
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Disruption Area
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        N/A
      </dd>
      <dd
        class="govuk-summary-list__actions"
      >
        <ul
          class="govuk-summary-list__actions-list"
        >
          <li
            class="govuk-summary-list__actions-list-item"
          >
            <a
              class="govuk-link"
              href="/create-consequence-network/2/0?return=%2Freview-disruption"
            >
              Change
              <span
                class="govuk-visually-hidden"
              >
                Disruption Area
              </span>
            </a>
          </li>
        </ul>
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Advice to display
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        The road is closed for the following reasons: Example, example, example, example
      </dd>
      <dd
        class="govuk-summary-list__actions"
      >
        <ul
          class="govuk-summary-list__actions-list"
        >
          <li
            class="govuk-summary-list__actions-list-item"
          >
            <a
              class="govuk-link"
              href="/create-consequence-network/2/0?return=%2Freview-disruption"
            >
              Change
              <span
                class="govuk-visually-hidden"
              >
                Advice to display
              </span>
            </a>
          </li>
        </ul>
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Remove from journey planner
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        Yes
      </dd>
      <dd
        class="govuk-summary-list__actions"
      >
        <ul
          class="govuk-summary-list__actions-list"
        >
          <li
            class="govuk-summary-list__actions-list-item"
          >
            <a
              class="govuk-link"
              href="/create-consequence-network/2/0?return=%2Freview-disruption"
            >
              Change
              <span
                class="govuk-visually-hidden"
              >
                Remove from journey planner
              </span>
            </a>
          </li>
        </ul>
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Disruption delay
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        33 minutes
      </dd>
      <dd
        class="govuk-summary-list__actions"
      >
        <ul
          class="govuk-summary-list__actions-list"
        >
          <li
            class="govuk-summary-list__actions-list-item"
          >
            <a
              class="govuk-link"
              href="/create-consequence-network/2/0?return=%2Freview-disruption"
            >
              Change
              <span
                class="govuk-visually-hidden"
              >
                Disruption delay
              </span>
            </a>
          </li>
        </ul>
      </dd>
    </div>
  </dl>
  <button
    class="govuk-button govuk-button--warning mt-4"
    data-module="govuk-button"
  >
    Delete consequence
  </button>
  <button
    class="govuk-button govuk-button--secondary mt-4 ml-4"
    data-module="govuk-button"
    formaction="/api/duplicate-consequence?consequenceId=0&return=/review-disruption"
  >
    Duplicate consequence
  </button>
</DocumentFragment>
`;

exports[`ReviewConsequenceTable > should render the table with data without change links 1`] = `
<DocumentFragment>
  <dl
    class="govuk-summary-list"
  >
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Consequence type
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        Network wide
      </dd>
      <dd
        class="govuk-summary-list__actions"
      >
        <ul
          class="govuk-summary-list__actions-list"
        />
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Mode of transport
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        Bus
      </dd>
      <dd
        class="govuk-summary-list__actions"
      >
        <ul
          class="govuk-summary-list__actions-list"
        />
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Disruption Area
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        N/A
      </dd>
      <dd
        class="govuk-summary-list__actions"
      >
        <ul
          class="govuk-summary-list__actions-list"
        />
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Advice to display
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        The road is closed for the following reasons: Example, example, example, example
      </dd>
      <dd
        class="govuk-summary-list__actions"
      >
        <ul
          class="govuk-summary-list__actions-list"
        />
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Remove from journey planner
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        Yes
      </dd>
      <dd
        class="govuk-summary-list__actions"
      >
        <ul
          class="govuk-summary-list__actions-list"
        />
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Disruption delay
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        33 minutes
      </dd>
      <dd
        class="govuk-summary-list__actions"
      >
        <ul
          class="govuk-summary-list__actions-list"
        />
      </dd>
    </div>
  </dl>
</DocumentFragment>
`;
